src/main/scala/zio/telemetry/Telemetry.scala

```diff
@@ -0,0 +1,213 @@
+          zio.catchAllCause { cause =>
+            tag("error", true).when(tagError) *>
+              log(
+                Map("error.object" -> cause, "stack" -> cause.prettyPrint)
+              ).when(logError) *>
+              IO.done(Exit.Failure(cause))
+          }
+      )
+
+  def context: ZIO[Telemetry, Nothing, SpanContext] =
+    getSpan.map(_.context)
+
+  def getBaggageItem(key: String): ZIO[Telemetry, Nothing, Option[String]] =
+    getSpan.map(_.getBaggageItem(key)).map(Option.apply)
+
+  def setBaggageItem(
+      key: String,
+      value: String
+  ): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setBaggageItem(key, value))).unit
+
+  def tag(key: String, value: String): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def tag(key: String, value: Int): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def tag(key: String, value: Boolean): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def tag[T <: Object](key: Tag[T], value: T): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def log(msg: String): ZIO[COT, Nothing, Unit] =
+    for {
+      span <- getSpan
+      now <- getCurrentTimeMicros
+      _ <- UIO(span.log(now, msg))
+    } yield ()
+
+  def log(fields: Map[String, _]): ZIO[COT, Nothing, Unit] =
+    for {
+      span <- getSpan
+      now <- getCurrentTimeMicros
+      _ <- UIO(span.log(now, fields.asJava))
+    } yield ()
+
+  private def getSpan: ZIO[Telemetry, Nothing, Span] =
+    ZIO.accessM[Telemetry](_.telemetry.spanRef.get)
+
+  private def setSpan(span: Span): ZIO[Telemetry, Nothing, Unit] =
+    ZIO.accessM[Telemetry](_.telemetry.spanRef.set(span))
+
+  private def getTracer: ZIO[Telemetry, Nothing, Tracer] =
+    ZIO.environment[Telemetry].map(_.telemetry.tracer)
+
+  private def getCurrentTimeMicros: ZIO[Clock, Nothing, Long] =
+    ZIO.accessM[Clock](_.clock.currentTime(TimeUnit.MICROSECONDS))
+
+  implicit class Implicits[R, E, A](zio: ZIO[R, E, A]) extends AnyRef {
+
+    def spanFrom[C <: Object](
+        format: Format[C],
+        carrier: C,
+        opName: String,
+        tagError: Boolean = true,
+        logError: Boolean = true
+    ): ZIO[R with COT, E, A] =
+      Telemetry.spanFrom(
+        format,
+        carrier,
+        zio,
+        opName,
+        tagError,
+        logError
+      )
+
+    def rootSpan(
+        opName: String,
+        tagError: Boolean = true,
+        logError: Boolean = true
+    ): ZIO[R with COT, E, A] =
+      Telemetry.rootSpan(zio, opName, tagError, logError)
+
+    def span(
+        opName: String,
+        tagError: Boolean = true,
+        logError: Boolean = true
+    ): ZIO[R with COT, E, A] =
+      Telemetry.span(zio, opName, tagError, logError)
+
+    def span(
+        span: Span,
+        tagError: Boolean,
```

Comment:
Perhaps add defaults here?

Reply:
We can only have one `span` method  with defaults.